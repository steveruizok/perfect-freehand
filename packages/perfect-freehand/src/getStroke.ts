import type { StrokeOptions } from './types'
import { getStrokeOutlinePoints } from './getStrokeOutlinePoints'
import { getStrokePoints } from './getStrokePoints'

/**
 * ## getStroke
 * @description Get an array of points describing a polygon that surrounds the input points.
 * @param points An array of points (as `[x, y, pressure]` or `{x, y, pressure}`). Pressure is optional in both cases.
 * @param options (optional) An object with options.
 * @param options.size	The base size (diameter) of the stroke.
 * @param options.thinning The effect of pressure on the stroke's size.
 * @param options.smoothing	How much to soften the stroke's edges.
 * @param options.easing	An easing function to apply to each point's pressure.
 * @param options.simulatePressure Whether to simulate pressure based on velocity.
 * @param options.start Tapering and easing function for the start of the line.
 * @param options.end Tapering and easing function for the end of the line.
 * @param options.last Whether to handle the points as a completed stroke.
 */
export function getStroke<T extends number[]>(
  points: T[],
  options?: StrokeOptions
): number[][]
export function getStroke<
  K extends { x: number; y: number; pressure?: number }
>(points: K[], options?: StrokeOptions): number[][]
export function getStroke<
  T extends number[],
  K extends { x: number; y: number; pressure?: number }
>(points: (T | K)[], options: StrokeOptions = {} as StrokeOptions): number[][] {
  return getStrokeOutlinePoints(getStrokePoints(points, options), options)
}
