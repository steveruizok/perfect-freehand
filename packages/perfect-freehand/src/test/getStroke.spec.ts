import { getStroke } from '../getStroke'
import inputs from './inputs.json'

const { onePoint, twoPoints, manyPoints, withDuplicates } = inputs

describe('getStroke', () => {
  it('runs on a line with no points', () => {
    expect(getStroke([])).toMatchSnapshot('get-stroke-no-points')
  })

  it('runs on a line with a single point', () => {
    expect(getStroke(onePoint)).toMatchSnapshot('get-stroke-one-point')
  })

  it('runs on a line with two points', () => {
    expect(getStroke(twoPoints)).toMatchSnapshot('get-stroke-one-point')
  })

  it('runs on a line with a many points', () => {
    expect(getStroke(manyPoints)).toMatchSnapshot('get-stroke-many-points')
  })

  it('runs on a line with withDuplicates', () => {
    expect(getStroke(withDuplicates)).toMatchSnapshot('get-stroke-one-point')
  })

  it('Caps points', () => {
    expect(getStroke(twoPoints).length > 4).toBeTruthy()
  })

  it('Succeeds on tricky points', () => {
    expect(JSON.stringify(getStroke(manyPoints)).includes('null')).toBeFalsy()
  })

  it('Solves a tricky stroke with only one point.', () => {
    const stroke = getStroke(onePoint, {
      size: 1,
      thinning: 0.6,
      smoothing: 0.5,
      streamline: 0.5,
      simulatePressure: true,
      last: false,
    })

    expect(stroke).toMatchSnapshot()

    expect(Number.isNaN([0][0])).toBe(false)
  })
})
