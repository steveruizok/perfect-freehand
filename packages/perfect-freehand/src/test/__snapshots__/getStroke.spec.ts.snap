// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStroke Solves a tricky stroke with only one point. 1`] = `
Array [
  Array [
    465.08677669529663,
    286.3332233047034,
  ],
  Array [
    465.1569220851488,
    286.4708913837399,
  ],
  Array [
    465.1327516310471,
    286.6234976249349,
  ],
  Array [
    465.0234976249349,
    286.73275163104705,
  ],
  Array [
    464.87089138373995,
    286.75692208514874,
  ],
  Array [
    464.7332233047034,
    286.6867766952966,
  ],
  Array [
    464.66307791485127,
    286.54910861626007,
  ],
  Array [
    464.68724836895296,
    286.3965023750651,
  ],
  Array [
    464.79650237506513,
    286.28724836895293,
  ],
  Array [
    464.94910861626005,
    286.26307791485124,
  ],
  Array [
    465.08677669529663,
    286.3332233047034,
  ],
]
`;

exports[`getStroke runs on a line with a many points: get-stroke-many-points 1`] = `
Array [
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.3090704738867701,
    -0.7683141170461836,
  ],
  Array [
    0.02320812565912045,
    -3.9570438217523636,
  ],
  Array [
    2.001042394381983,
    -6.424724312433344,
  ],
  Array [
    5.088066183072562,
    -9.936857912484465,
  ],
  Array [
    8.357894303288951,
    -14.206135580792939,
  ],
  Array [
    12.893500261333543,
    -20.483413096976378,
  ],
  Array [
    19.2522016754106,
    -29.758191111094394,
  ],
  Array [
    26.088586813749092,
    -39.21977961655499,
  ],
  Array [
    33.001913433466406,
    -49.6726071577158,
  ],
  Array [
    40.07883355616558,
    -61.058475228055215,
  ],
  Array [
    47.565801007637944,
    -73.40264620889145,
  ],
  Array [
    55.315952242337005,
    -85.90323757834877,
  ],
  Array [
    63.08401611782964,
    -98.60894004361714,
  ],
  Array [
    70.02638514953443,
    -109.59504332899755,
  ],
  Array [
    75.24893818075917,
    -118.02973149533473,
  ],
  Array [
    80.4535701046435,
    -126.2915493903768,
  ],
  Array [
    85.1475895355663,
    -133.89526586905365,
  ],
  Array [
    89.6319517052493,
    -140.70834046478802,
  ],
  Array [
    94.10093698005623,
    -146.84284689826532,
  ],
  Array [
    98.1048699260494,
    -153.00576586701877,
  ],
  Array [
    102.31194753076497,
    -159.6295059579491,
  ],
  Array [
    106.10223366189906,
    -165.65156177517457,
  ],
  Array [
    109.73408603737155,
    -170.94917950062424,
  ],
  Array [
    112.73331694347478,
    -175.06607520270435,
  ],
  Array [
    115.52755958742551,
    -179.01043239220107,
  ],
  Array [
    118.33316080245652,
    -183.3833588152735,
  ],
  Array [
    122.0312730033545,
    -188.87898862679086,
  ],
  Array [
    125.13029023133312,
    -193.0051440040649,
  ],
  Array [
    126.17223890831708,
    -193.45165467873142,
  ],
  Array [
    127.27764548482374,
    -193.2004473367296,
  ],
  Array [
    128.0242822200152,
    -192.34747464448068,
  ],
  Array [
    128.1269592582976,
    -191.2185431789482,
  ],
  Array [
    128.21800292173742,
    -189.31090570757635,
  ],
  Array [
    128.48572306801339,
    -182.2846662182831,
  ],
  Array [
    128.43646935439006,
    -176.6476511151247,
  ],
  Array [
    128.28857155986566,
    -169.27796049981615,
  ],
  Array [
    127.89741897888021,
    -160.01562069715854,
  ],
  Array [
    126.47585365535811,
    -147.59848061037695,
  ],
  Array [
    123.76341994859209,
    -131.02687759146045,
  ],
  Array [
    121.01839041793136,
    -116.45044589143603,
  ],
  Array [
    117.84026586368589,
    -100.92016164445789,
  ],
  Array [
    114.59356395659844,
    -84.09811148079137,
  ],
  Array [
    111.01888361868677,
    -66.14067697503387,
  ],
  Array [
    107.23429066158172,
    -48.37512788733251,
  ],
  Array [
    103.44535543431758,
    -30.846906072957317,
  ],
  Array [
    100.19490812913605,
    -14.618534654402856,
  ],
  Array [
    97.82671395235187,
    0.12818570210473634,
  ],
  Array [
    95.68100012794257,
    14.042109662690454,
  ],
  Array [
    93.71899443677454,
    27.29175382237143,
  ],
  Array [
    92.28019218797715,
    39.31390698101868,
  ],
  Array [
    91.38135120880115,
    49.95865419880831,
  ],
  Array [
    90.5873622684783,
    59.81980588661869,
  ],
  Array [
    90.1489528087084,
    68.70496944718272,
  ],
  Array [
    89.97838700029845,
    75.58256458351224,
  ],
  Array [
    89.95021193429477,
    81.60819213363806,
  ],
  Array [
    89.96792207626123,
    87.44518590113488,
  ],
  Array [
    90.01800783456915,
    92.7099533357507,
  ],
  Array [
    90.11142903853012,
    97.6205425031496,
  ],
  Array [
    90.25753015916257,
    103.38258901790455,
  ],
  Array [
    90.40122147212801,
    107.68789535096326,
  ],
  Array [
    90.05605978572709,
    108.74889983023094,
  ],
  Array [
    89.15317533009183,
    109.40438953614357,
  ],
  Array [
    88.03743927936095,
    109.4039896803182,
  ],
  Array [
    87.13502488244008,
    108.74785299408953,
  ],
  Array [
    86.83641608818127,
    106.86426363086727,
  ],
  Array [
    85.52184305935684,
    99.73322513248914,
  ],
  Array [
    83.71646973873678,
    93.42854144588279,
  ],
  Array [
    81.43353458369329,
    85.77755276680779,
  ],
  Array [
    78.63929759084265,
    77.25041778528178,
  ],
  Array [
    75.0171856626504,
    67.43198915576941,
  ],
  Array [
    70.85687352652305,
    56.03087471969231,
  ],
  Array [
    65.7119701975102,
    43.78828134893365,
  ],
  Array [
    60.5441300940441,
    31.967126076543654,
  ],
  Array [
    55.34729233951133,
    20.42192553230031,
  ],
  Array [
    50.025553663510635,
    8.813423802110457,
  ],
  Array [
    45.233687311223356,
    -2.367986257445635,
  ],
  Array [
    40.82817539189043,
    -12.477916536835934,
  ],
  Array [
    36.628846518036646,
    -21.71544683816006,
  ],
  Array [
    32.62908081836294,
    -30.586083252847853,
  ],
  Array [
    29.152790967775992,
    -38.3271108103946,
  ],
  Array [
    26.283476538282827,
    -44.37551741423302,
  ],
  Array [
    23.77759081057429,
    -49.5159309989945,
  ],
  Array [
    21.293017635198083,
    -54.49982853402538,
  ],
  Array [
    18.52156441381262,
    -59.06010417311025,
  ],
  Array [
    15.966550232496187,
    -63.06507486282652,
  ],
  Array [
    12.92040132604382,
    -67.61530062631566,
  ],
  Array [
    10.842148126604055,
    -70.5067456765801,
  ],
  Array [
    10.68717323115917,
    -72.10813401377621,
  ],
  Array [
    11.891631652526595,
    -73.17478378568964,
  ],
  Array [
    13.462532342129972,
    -72.82731745907722,
  ],
  Array [
    14.104778437092428,
    -71.35219652226175,
  ],
  Array [
    13.288883313411779,
    -69.96555418485818,
  ],
  Array [
    11.687494976215664,
    -69.8105792894133,
  ],
  Array [
    10.620845204302226,
    -71.01503771078072,
  ],
  Array [
    10.968311530914646,
    -72.5859384003841,
  ],
  Array [
    12.443432467730123,
    -73.22818449534655,
  ],
  Array [
    13.83007480513369,
    -72.4122893716659,
  ],
  Array [
    18.582040167923008,
    -64.59382925685686,
  ],
  Array [
    20.945566437312653,
    -60.487860524344505,
  ],
  Array [
    23.70493841266496,
    -55.660124653636075,
  ],
  Array [
    26.058367793609023,
    -50.578063785014095,
  ],
  Array [
    28.386600155609152,
    -45.34694825807245,
  ],
  Array [
    31.13419976794228,
    -39.21462281212001,
  ],
  Array [
    34.58380517546171,
    -31.468143311777414,
  ],
  Array [
    38.588233038723494,
    -22.60750786084065,
  ],
  Array [
    42.81170957107713,
    -13.344947931840299,
  ],
  Array [
    47.23831814617766,
    -3.229418756129511,
  ],
  Array [
    52.02975087486904,
    7.891764061229841,
  ],
  Array [
    57.388958913496,
    19.498491885150184,
  ],
  Array [
    62.64010590424067,
    31.044013250802934,
  ],
  Array [
    67.8847831874623,
    42.86587860825937,
  ],
  Array [
    73.1642020086222,
    55.17878735751906,
  ],
  Array [
    77.44825843362378,
    66.50544232343151,
  ],
  Array [
    81.28922919238988,
    76.34314346961452,
  ],
  Array [
    84.36723096165892,
    84.83181386977756,
  ],
  Array [
    87.00521366007288,
    92.43371093587137,
  ],
  Array [
    89.10031802167639,
    98.96244863034681,
  ],
  Array [
    89.93813317902261,
    106.691873397679,
  ],
  Array [
    90.05605978572709,
    108.74889983023094,
  ],
  Array [
    89.15317533009183,
    109.40438953614357,
  ],
  Array [
    88.03743927936095,
    109.4039896803182,
  ],
  Array [
    87.13502488244008,
    108.74785299408953,
  ],
  Array [
    86.79062376701575,
    107.68660139033116,
  ],
  Array [
    86.9407746939288,
    103.37775482903068,
  ],
  Array [
    87.10813747215803,
    97.61069096431673,
  ],
  Array [
    87.23384953065177,
    92.6943739916162,
  ],
  Array [
    87.34418131449632,
    87.41853099894517,
  ],
  Array [
    87.48220724389657,
    81.55966948642667,
  ],
  Array [
    87.69360617092352,
    75.50577992387174,
  ],
  Array [
    87.99584285755353,
    68.59750151893942,
  ],
  Array [
    88.46496544173583,
    59.64988210157732,
  ],
  Array [
    89.2576029125504,
    49.77925246112042,
  ],
  Array [
    90.16138869504783,
    39.06314478788445,
  ],
  Array [
    91.60408360598892,
    26.9798347575348,
  ],
  Array [
    93.56253405264745,
    13.716032097547174,
  ],
  Array [
    95.70250341398864,
    -0.21206023125249374,
  ],
  Array [
    98.07430106928982,
    -15.040108353804891,
  ],
  Array [
    101.31559019178542,
    -31.305689295105473,
  ],
  Array [
    105.0806444009142,
    -48.831864610851056,
  ],
  Array [
    108.8271298028273,
    -66.57406424585501,
  ],
  Array [
    112.35320310034541,
    -84.52585396029066,
  ],
  Array [
    115.5395899684789,
    -101.38461288570548,
  ],
  Array [
    118.61713199084285,
    -116.8901107090603,
  ],
  Array [
    121.22036977040659,
    -131.43022650272343,
  ],
  Array [
    123.72607562612262,
    -147.86161858931212,
  ],
  Array [
    124.87041907976342,
    -160.06986942232095,
  ],
  Array [
    124.88886196629728,
    -169.27471571192433,
  ],
  Array [
    124.72682258597214,
    -176.63782279721886,
  ],
  Array [
    124.64070813179993,
    -185.0626362995509,
  ],
  Array [
    125.78526847616719,
    -191.9127076251737,
  ],
  Array [
    126.17223890831708,
    -193.45165467873142,
  ],
  Array [
    127.27764548482374,
    -193.2004473367296,
  ],
  Array [
    128.0242822200152,
    -192.34747464448068,
  ],
  Array [
    128.1269592582976,
    -191.2185431789482,
  ],
  Array [
    127.54645746091121,
    -190.24486638899634,
  ],
  Array [
    124.57729755656494,
    -186.7748562311013,
  ],
  Array [
    120.88260738455844,
    -181.70854396013956,
  ],
  Array [
    117.82627048408528,
    -177.32394834617628,
  ],
  Array [
    114.86320227035495,
    -173.44617662936957,
  ],
  Array [
    111.74998724991863,
    -169.5262119476129,
  ],
  Array [
    108.070002505248,
    -164.3944934780086,
  ],
  Array [
    104.21494279371187,
    -158.4241843408442,
  ],
  Array [
    100.00129909878298,
    -151.7610391373028,
  ],
  Array [
    95.88361175966301,
    -145.5372552029611,
  ],
  Array [
    91.42829034929835,
    -139.5207619453382,
  ],
  Array [
    86.97764245896674,
    -132.7662738494136,
  ],
  Array [
    82.28326163183101,
    -125.14148508802084,
  ],
  Array [
    77.10081221142262,
    -116.88601336241297,
  ],
  Array [
    71.88751758500774,
    -108.42444744438608,
  ],
  Array [
    64.98808148710845,
    -97.4479935026537,
  ],
  Array [
    57.25187621744385,
    -84.7082871633884,
  ],
  Array [
    49.569535846134315,
    -72.19629482192923,
  ],
  Array [
    42.1478001914246,
    -59.78470176983754,
  ],
  Array [
    35.14499473677003,
    -48.275462107751544,
  ],
  Array [
    28.32495401184248,
    -37.63659618464436,
  ],
  Array [
    21.751107321267202,
    -28.10592976506595,
  ],
  Array [
    15.628362239772336,
    -18.542989950088476,
  ],
  Array [
    11.235178180206503,
    -12.025583933975703,
  ],
  Array [
    7.8520441064026905,
    -7.490007293060183,
  ],
  Array [
    4.860850554256603,
    -4.09950880687392,
  ],
  Array [
    3.048492424528695,
    -2.0043843183441687,
  ],
  Array [
    0.4570507507838458,
    -0.3058020231258438,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.28631680253626735,
    0.11404551630438202,
  ],
  Array [
    -0.19072914258565155,
    0.5216825340665945,
  ],
  Array [
    -0.32761802548502206,
    0.5036052924641046,
  ],
  Array [
    -0.45222091470824144,
    0.44411182288450696,
  ],
  Array [
    -0.5523408112987629,
    0.3490257606456294,
  ],
  Array [
    -0.618177282201968,
    0.22765479202296496,
  ],
]
`;

exports[`getStroke runs on a line with a single point: get-stroke-one-point 1`] = `
Array [
  Array [
    466.3338465858493,
    285.0861534141507,
  ],
  Array [
    466.89883211002405,
    286.1949999393544,
  ],
  Array [
    466.70415136608676,
    287.42416578126637,
  ],
  Array [
    465.8241657812664,
    288.3041513660867,
  ],
  Array [
    464.5949999393544,
    288.498832110024,
  ],
  Array [
    463.48615341415075,
    287.93384658584927,
  ],
  Array [
    462.921167889976,
    286.8250000606456,
  ],
  Array [
    463.1158486339133,
    285.5958342187336,
  ],
  Array [
    463.99583421873365,
    284.71584863391325,
  ],
  Array [
    465.22500006064564,
    284.52116788997597,
  ],
  Array [
    466.3338465858493,
    285.0861534141507,
  ],
]
`;

exports[`getStroke runs on a line with no points: get-stroke-no-points 1`] = `Array []`;

exports[`getStroke runs on a line with two points: get-stroke-one-point 1`] = `
Array [
  Array [
    8,
    200,
  ],
  Array [
    8.381966011250105,
    198.82442949541505,
  ],
  Array [
    9.381966011250105,
    198.0978869674097,
  ],
  Array [
    10.618033988749895,
    198.0978869674097,
  ],
  Array [
    11.618033988749895,
    198.82442949541505,
  ],
  Array [
    12,
    200,
  ],
  Array [
    11.618033988749895,
    201.17557050458495,
  ],
  Array [
    10.618033988749895,
    201.9021130325903,
  ],
  Array [
    9.381966011250105,
    201.9021130325903,
  ],
  Array [
    8.381966011250107,
    201.17557050458495,
  ],
  Array [
    8,
    200,
  ],
]
`;

exports[`getStroke runs on a line with withDuplicates: get-stroke-one-point 1`] = `
Array [
  Array [
    3.263339185919482,
    0.9033274807471852,
  ],
  Array [
    28.566457301185352,
    26.433435531462106,
  ],
  Array [
    89.13276324824002,
    87.11689738847693,
  ],
  Array [
    101.04044751235682,
    98.95569440782458,
  ],
  Array [
    101.45406487016946,
    99.76746417952694,
  ],
  Array [
    101.31154204174487,
    100.66731790336847,
  ],
  Array [
    100.66731790336847,
    101.31154204174487,
  ],
  Array [
    99.76746417952694,
    101.45406487016946,
  ],
  Array [
    78.75310214840889,
    80.49621721887097,
  ],
  Array [
    1.821395088375465,
    3.7340532982962378,
  ],
  Array [
    1.4419215509737402,
    2.5661538394412915,
  ],
  Array [
    2.1637231119887064,
    1.5726792206924218,
  ],
  Array [
    3.391725274682997,
    1.572679220692422,
  ],
  Array [
    4.113526835697963,
    2.5661538394412915,
  ],
  Array [
    3.7340532982962378,
    3.7340532982962378,
  ],
  Array [
    2.566153839441292,
    4.113526835697963,
  ],
  Array [
    1.572679220692422,
    3.391725274682997,
  ],
  Array [
    1.5726792206924212,
    2.1637231119887077,
  ],
  Array [
    2.5661538394412915,
    1.4419215509737402,
  ],
  Array [
    3.734053298296236,
    1.821395088375463,
  ],
  Array [
    80.49621721887097,
    78.75310214840889,
  ],
  Array [
    101.45406487016946,
    99.76746417952694,
  ],
  Array [
    101.31154204174487,
    100.66731790336847,
  ],
  Array [
    100.66731790336847,
    101.31154204174487,
  ],
  Array [
    99.76746417952694,
    101.45406487016946,
  ],
  Array [
    98.95569440782458,
    101.04044751235682,
  ],
  Array [
    87.11689738847693,
    89.13276324824002,
  ],
  Array [
    26.433435531462106,
    28.566457301185352,
  ],
  Array [
    0.9033274807471852,
    3.263339185919482,
  ],
  Array [
    -1.1800058525861483,
    1.1800058525861483,
  ],
  Array [
    -1.4868941172794934,
    0.7576103934574558,
  ],
  Array [
    -1.648234825973037,
    0.26105475043514903,
  ],
  Array [
    -1.648234825973037,
    -0.26105475043514914,
  ],
  Array [
    -1.4868941172794936,
    -0.7576103934574557,
  ],
  Array [
    -1.1800058525861483,
    -1.1800058525861483,
  ],
  Array [
    -0.7576103934574561,
    -1.4868941172794936,
  ],
  Array [
    -0.26105475043514914,
    -1.648234825973037,
  ],
  Array [
    0.2610547504351487,
    -1.648234825973037,
  ],
  Array [
    0.7576103934574552,
    -1.486894117279494,
  ],
  Array [
    1.1800058525861477,
    -1.180005852586149,
  ],
]
`;
